packages/core/src/application.ts

```diff
@@ -525,6 +525,9 @@ export interface ApplicationOptions {
   http?: HttpConfig;
   components?: Array<Constructor<Component>>;
   sequence?: Constructor<SequenceHandler>;
+
+  // tslint:disable-next-line:no-any
+  [prop: string]: any;
```

Comment:
Is there a way how to avoid this? My concern is that by allowing any property with any value, we are loosing type safety.

Reply:
Just double-checking my TypeScript knowledge here, but doesn't this allow for JavaScript users to duck-type their own implementations of ApplicationOptions so they don't have to explicitly inherit/extend it as they work, or is it not quite that smart?

The other interesting thing I'm seeing is what's _not_ part of the PR, which is the greater problem of having ApplicationOptions contain an optional HttpConfig instance.

I think at its minimum, this interface should only define the components and sequence handler, and they should _not_ be optional; there should always be a sequence handler, and (an empty) array of Components.

```ts
export interface ApplicationOptions {
  components: Array<Constructor<Component>>;
  sequence: Constructor<SequenceHandler>;
  // tslint:disable-next-line:no-any
  [prop: string]: any;
}
```

Reply:
To be clear, I don't think this should happen as a part of this PR, it would be a part of the greater rework to break defaults out of the Application definition.

Reply:
> Just double-checking my TypeScript knowledge here, but doesn't this allow for JavaScript users to duck-type their own implementations of ApplicationOptions so they don't have to explicitly inherit/extend it as they work, or is it not quite that smart?

Application `ApplicationOptions` is defined as an interface. TypeScript checks interfaces by comparing the structure of the object, therefore you can use duck-typing in TypeScript too, no need to inherit from this interface.

> The other interesting thing I'm seeing is what's not part of the PR, which is the greater problem of having ApplicationOptions contain an optional HttpConfig instance.

+1, I think we should move all things REST-related from `core` into a standalone package/extension.

> I think at its minimum, this interface should only define the components and sequence handler, and they should not be optional; there should always be a sequence handler, and (an empty) array of Components.

I personally think they should be both optional. We have a default sequence that's good enough for simple (starter) applications. I also don't see why should we force users to configure an empty array of components, when it's easier to simply omit the property at all?

BTW depending on how the gRPC spike turns out, we may need to move sequence into the REST-specific component too, in which case it should be removed from core application options (similarly like repositories are not included by default).

> To be clear, I don't think this should happen as a part of this PR, it would be a part of the greater rework to break defaults out of the Application definition.

+1

Reply:
To simplify things, I think we should consider to have a default component for each application or have `Application` extends from `Component`. Then the component can be universal container for all extensions.

Reply:
I like the idea, although I think it's `ApplicationOptions` that should have the same interface as the `Component` instance?

Reply:
Same reason as above. We can add repositories as a first class citizen but importing `Repository` type from `@loopback/repository` will create a dependency & creating an empty type seems pointless. Also I think it's an important to allow extensions developers to be able to specify their own properties. This also allows them to call `super()` ... required in the mixin constructor.